src/pqm/pqm.py

```diff
@@ -1,35 +1,119 @@
+    sy = torch.std(sample2, dim=dim, unbiased=True)
     m = (n1 * mx + n2 * my) / (n1 + n2)
-    s = np.sqrt(
-        ((n1 - 1) * (sx**2) + (n2 - 1) * (sy**2) + n1 * n2 * (mx - my) ** 2 / (n1 + n2))
+    s = torch.sqrt(
+        (
+            (n1 - 1) * (sx ** 2)
+            + (n2 - 1) * (sy ** 2)
+            + n1 * n2 * (mx - my) ** 2 / (n1 + n2)
+        )
         / (n1 + n2 - 1)
     )
     return m, s
 
 
+def rescale_chi2(chi2_stat, orig_dof, target_dof, device):
+    """
+    Rescale chi2 statistic using appropriate methods depending on the device.
+    """        
+
+    # Move tensors to CPU and convert to NumPy
+    chi2_stat_cpu = chi2_stat.cpu().item()  # Convert to float
+    orig_dof_cpu = orig_dof.cpu().item()    # Convert to float
+
+    if orig_dof_cpu == target_dof:
+        return chi2_stat_cpu
+        
+    if chi2_stat_cpu / orig_dof_cpu < 10:
+        # Use cumulative probability method for better accuracy
+        cp = chi2.sf(chi2_stat_cpu, orig_dof_cpu)
+        return chi2.isf(cp, target_dof)
+    else:
+        # Use simple scaling for large values
+        return chi2_stat_cpu * target_dof / orig_dof_cpu
+
+
+
+def _chi2_contingency(counts, device):
+    """
+    Computes the chi-squared statistic and p-value for a contingency table.
+
+    Parameters
+    ----------
+    counts: torch.Tensor
+        2xN tensor of counts for each category.
+    device : str
+        Device to use for computation. Default: 'cpu'. If 'cuda' is selected,
+
+    Returns
+    -------
+    tuple
+        chi2_stat, p_value, dof, expected
+    """
+    if device == 'cpu':
+        counts_np = counts.cpu().numpy()
+        chi2_stat, p_value, dof, expected = chi2_contingency(counts_np)
+        chi2_stat = torch.tensor(chi2_stat, device=device)
+        dof = torch.tensor(dof, device=device)
+        return chi2_stat, p_value, dof, expected
+    else:
+        # Observed counts
+        O = counts.float()
+
+        # Row sums and column sums
+        row_sums = O.sum(dim=1, keepdim=True)  # shape (2, 1)
+        col_sums = O.sum(dim=0, keepdim=True)  # shape (1, N)
+        total = O.sum()
+
+        # Expected counts under the null hypothesis of independence
+        E = row_sums @ col_sums / total  # shape (2, N)
+
+        # Degrees of freedom
+        dof = (O.size(0) - 1) * (O.size(1) - 1)
+
+        # Avoid division by zero
+        mask = E > 0
+        O = O[mask]
+        E = E[mask]
+
+        # Compute chi-squared statistic
+        chi2_stat = ((O - E) ** 2 / E).sum()
+
+        # Move dof and chi2_stat to the same device
+        dof = torch.tensor(dof, dtype=torch.float32, device=chi2_stat.device)
+
+        # Compute p-value using the survival function (1 - CDF)
+        p_value = torch.special.gammaincc(dof / 2, chi2_stat / 2).item()
+
+        return chi2_stat, p_value, dof, E
+
+
 def _pqm_test(
-    x_samples: np.ndarray,
-    y_samples: np.ndarray,
+    x_samples: torch.Tensor,
```

Comment:
We should definitely keep numpy input as an option since it is less aggressive than pytorch about requiring everything to be arrays. Numpy is a good common ground say for people who use JAX that still want to use PQM. I'm sure JAX has a to_numpy function, but not a to_torch option.